send_queue.go

```diff
@@ -1,29 +1,55 @@
 package quic
 
+type sender interface {
+	Send(p *packetBuffer)
+	Run() error
+	WouldBlock() bool
+	Available() <-chan struct{}
+	Close()
+}
+
 type sendQueue struct {
 	queue       chan *packetBuffer
 	closeCalled chan struct{} // runStopped when Close() is called
 	runStopped  chan struct{} // runStopped when the run loop returns
+	available   chan struct{}
```

Comment:
This is a bit difficult to follow, and tightly coupled to run loop internals (e.g. if you use it wrong it's inherently racy, see below).

I wonder whether there's a simpler way of achieving similar behavior? Not sure if it's simpler, but possibly a nextPacket member in the session that we try to send in the session's switch statement?

(If we decide to stick with this, we should probably document the reasoning behind this API.)

Reply:
I'm not sure I understand why it's inherently race. It's true that the `sendQueue` can only be used from a single Go routine (which we guarantee by just running it from the session), but as long as you adhere to this, there shouldn't be any race here.